exclude all processes after self (fix bug #24)

lesmana opened the pull request:
fixes bug #24 
replaces pull request #27

lesmana committed:
exclude all processes after self

walles commented:
There's a corner case here that doesn't work.

Let's call the processes involved `px`, `ps.sh` and `ps`.

As long as `ps` returns them all in the right order, everything will be fine.

But if they arrive as `px`, `ps` and `ps.sh`, then `ps`'s parent PID won't be in the `exclude` set when it is checked, and the check will fail.

I think you first need multiple passes to build the `exclude` set, then you can use it to filter out all processes in it.

Do you see the problem?

lesmana commented:
i see the problem. you are correct that it will fail if the processes come in a different order.

i have been thinking about it. i will implement a better solution and push that.

lesmana committed:
remove own and descendants after having all processes in a tree

lesmana commented:
i rebased my branch a couple of times to include your updates. it seems that it triggered some bots after every rebase. sorry about that.

i moved the pruning to the part of code where the process tree is already completed. i used DFS (or BFS, depends on pythons semantic of `append()` and `pop()`) to remove all descendants. the repeated `remove()` might be costly. but i think that is negligible for the typical use case of two processes to remove.

the removing might be done with a set difference operation
https://docs.python.org/2/library/stdtypes.html#set.difference
which should be more efficient than repeated `remove()`. For that `processes` needs to be a `set`. Which is also semantically correct since each process should be unique in the collection.

from a clean code perspective i would rather make an extra function for the pruning. but i needed the `pid2process` table. maybe the code can be restructured to expose that information in an elegant manner. i will think about that.

walles commented:
Don't worry about the bots, this is what they are there for.

If you run `test.sh` before pushing you'll improve your chances of them giving you a green light.

I hope your dev env works now, thanks for reporting all the issues @lesmana!

lesmana committed:
rename to processes

lesmana committed:
processes dict in get_all

lesmana committed:
extract remove own process

lesmana committed:
remove process takes pid as argument

lesmana commented:
i refactored a bit and made `processes` a dict with pids as keys. the return value from `get_all()` is still a list as before.

exposing the `processes` dict to the rest of the program might be advantageous. at least `find_process_by_pid()` in `px_processinfo.py` will benefit. 
https://github.com/walles/px/blob/b1731c7f3359bcfda5e3f4a33d53c23befd337cd/px/px_processinfo.py#L13

all other code need minor adjustments. for example
```
for process in processes.values():
```
instead
```
for process in processes:
```

lesmana committed:
conform flake8

walles commented on px/px_process.py:
```diff
@@ -222,46 +222,42 @@ def resolve_links(processes, now):
+    while toexclude:
+        process = toexclude.pop()
+        del processes[process.pid]
+        for child in process.children:
```
After this, won't `process.parent.children` still contain a reference to `process`?

lesmana replied:
yes you are correct. i fixed that. i opted to just remove the first reference because all other processes will be gone anyway.

walles commented:
I really like this approach here @lesmana.

How would you feel about adding (a) test(s) for your new `remove_process_and_descendants` method?

I'd be OK with merging this without tests, but I would prefer with, so just tell me which way you want to go.

lesmana committed:
remove process from parent

lesmana committed:
assert all parents and children in processes (no orphans)

walles commented on px/px_process.py:
```diff
@@ -222,46 +222,44 @@ def resolve_links(processes, now):
 
+def remove_process_and_descendants(processes, pid):
+    process = processes[pid]
+    process.parent.children.remove(process)
```
Do we need to check that the process parent isn't `None`?

But since only PID 0 has no parent and we're unlikely to want to remove that I think this is fine as it is.

Yay for CI, without that I'd just have guessed this fails all the time :).

lesmana replied:
i think it is fine without that check. if parent is ever none then something else before that already went way wrong. then again that check is not expensive and saves a lot of wondering and debugging later. i say you decide.

lesmana commented:
i studied your tests. i think the helper function `_validate_references()` already covers most of `remove_process_and_descendants()`.
https://github.com/walles/px/blob/b1731c7f3359bcfda5e3f4a33d53c23befd337cd/tests/px_process_test.py#L101

i added two asserts to that helper. asserting that all parents and children are also in `processes`. if you want to be really thorough we can also assert that every process in `processes` has at least one parent or child in `processes`. i leave that to you to decide whether that is overkill or not.

i can write more tests if you want. i prefer that you merge this as it is so we can close the bug. then we can refactor the tests somehwat to write more fine tuned tests.

some comments about the tests:

i see that your test style is more "integration" or "acceptance" level. testing a function along with all it's called functions. and that with live data from running system.

my preference for writing tests is different. i like to test every function as isolated as possible from other functions or methods. ideally testing only one function at a time. giving literals as input and asserting against literals. to that end i also like to write many small functions and compose them together.

for example i would write `get_all()` like this:
```
def get_all():
    ps_lines = call_ps()
    now = datetime.datetime.now().replace(tzinfo=dateutil.tz.tzlocal())
    processes = create_processes_dict(ps_lines)
    resolve_links(processes, now)
    remove_process_and_descendants(processes, os.getpid())
    return processes.values()
```
then write tests:
* `create_processes_dict()` by providing a hardcoded list containing string literals and asserting a dict.
* `resolve_links()` by providing a hardcoded dict and asserting a dict with another structure.
* `remove_process_and_descendants()` again by providing a hardcoded dict and asserting a dict with less processes in it.

and then just one "integration" level test for `get_all()` where i don't test all the edge cases because they are already covered in the other tests.

walles commented:
This is good; thank you for your persistence and good work @lesmana!

walles commented:
And you are right about how `get_all()` is structured and how to improve the testing of it.

By looking at the coverage data, the file with the highest amount of untested code is `px_processinfo.py`:
<https://coveralls.io/builds/12093752/source?filename=px%2Fpx_processinfo.py>

By the same reasoning you have, those `print_xxx()` methods should be rewritten into methods returning some lines to print, and then have tests written for them.

Thanks for the suggestion @lesmana!

walles commented:
Just released 1.0.7 with this code included. Thanks @lesmana!
